src/zenml/orchestrators/base_orchestrator.py

```diff
@@ -11,35 +11,410 @@
+
+from zenml.enums import MetadataContextTypes, StackComponentType
+from zenml.exceptions import DuplicateRunNameError
+from zenml.logger import get_logger
+from zenml.orchestrators import context_utils
+from zenml.orchestrators.utils import (
+    create_tfx_pipeline,
+    get_cache_status,
+    get_step_for_node,
+)
+from zenml.repository import Repository
 from zenml.stack import StackComponent
+from zenml.steps import BaseStep
+from zenml.steps.utils import (
+    INTERNAL_EXECUTION_PARAMETER_PREFIX,
+    PARAM_PIPELINE_PARAMETER_NAME,
+)
+from zenml.utils import string_utils
 
 if TYPE_CHECKING:
     from zenml.pipelines import BasePipeline
     from zenml.runtime_configuration import RuntimeConfiguration
     from zenml.stack import Stack
 
+logger = get_logger(__name__)
+
 
 class BaseOrchestrator(StackComponent, ABC):
-    """Base class for all ZenML orchestrators."""
+    """
+    Base class for all orchestrators. In order to implement an
+    orchestrator you will need to subclass from this class.
+
+    How it works:
+        The `run()` method is the entrypoint that is executed when the
+        pipeline's run method is called within the user code
+        (`pipeline_instance.run()`).
+
+        This method will take the ZenML Pipeline instance and prepare it for
+        eventual execution. To do this the following steps are taken:
+
+        * The underlying protobuf pipeline is created.
+
+        * Within the `configure_step_context()` method the pipeline
+        requirements, stack and runtime configuration is added to the step
+        context
+
+        * The `get_sorted_steps()` method then generates a sorted list of
+        steps which will later be used to directly execute these steps in order,
+        or to easily build a dag
+
+        * After these initial steps comes the most crucial one. Within the
+        `prepare_or_run_pipeline()` method each orchestrator will have its own
+        implementation that dictates the pipeline orchestration. In the simplest
+        case this method will iterate through all steps and execute them one by
+        one. In other cases this method will build and deploy an intermediate
+        representation of the pipeline (e.g an airflow dag or a kubeflow
+        pipelines yaml) to be executed within the orchestrators environment.
+
+    Building your own:
+        In order to build your own orchestrator, all you need to do is subclass
+        from this class and implement your own `prepare_or_run_pipeline()`
+        method. Overwriting other methods is NOT recommended but possible.
+        See the docstring of the prepare_or_run_pipeline method to find out
+        details of what needs to be implemented within it.
+    """
 
     # Class Configuration
     TYPE: ClassVar[StackComponentType] = StackComponentType.ORCHESTRATOR
 
     @abstractmethod
-    def run_pipeline(
+    def prepare_or_run_pipeline(
         self,
+        sorted_list_of_steps: List[BaseStep],
```

Comment:
What do you think of renaming this to `sorted_steps`, as the type is already visible from the type annotation (and to be honest doesn't really matter as long as you can loop over it)?